In [1]:
import fastervit

In [2]:
import torch
print(torch.__version__)

2.2.1+cu121


In [3]:
import torchvision
print(torchvision.__version__)

0.17.1+cu121


In [4]:
import timm
print(timm.__version__)

0.9.0


In [5]:
import sys
print(sys.version)

3.10.4 (main, Feb 17 2024, 10:59:11) [GCC 11.3.0]


In [8]:
import torch
from torchvision import transforms
from PIL import Image
from fastervit import create_model

# Define fastervit-0 model with 224 x 224 resolution

# Step 1: Load the pre-trained FasterViT model
model = create_model('faster_vit_0_224', 
                          pretrained=True,
                          model_path="/tmp/faster_vit_0.pth.tar")
model.eval()

FasterViT(
  (patch_embed): PatchEmbed(
    (proj): Identity()
    (conv_down): Sequential(
      (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (1): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (2): ReLU()
      (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (4): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (5): ReLU()
    )
  )
  (levels): ModuleList(
    (0): FasterViTLayer(
      (blocks): ModuleList(
        (0): ConvBlock(
          (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          (act1): GELU(approximate='none')
          (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_s

In [11]:
num_ftrs = model.head.in_features
model.head = torch.nn.Linear(num_ftrs, 8)

In [12]:
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
model = model.to(device)

In [13]:
model

FasterViT(
  (patch_embed): PatchEmbed(
    (proj): Identity()
    (conv_down): Sequential(
      (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (1): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (2): ReLU()
      (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (4): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (5): ReLU()
    )
  )
  (levels): ModuleList(
    (0): FasterViTLayer(
      (blocks): ModuleList(
        (0): ConvBlock(
          (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          (act1): GELU(approximate='none')
          (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_s

In [14]:
device

device(type='cpu')

In [15]:
import torchvision

print(torchvision.__version__)

0.17.1+cu121


In [17]:
model.state_dict()

OrderedDict([('patch_embed.conv_down.0.weight',
              tensor([[[[ 9.8405e-03,  7.9137e-04,  1.3553e-02],
                        [-2.9344e-03, -1.7916e-02, -7.7805e-03],
                        [ 5.1475e-03, -1.0829e-02,  2.4744e-03]],
              
                       [[-5.4770e-02, -6.9579e-02, -6.0724e-02],
                        [-7.7647e-02, -1.0245e-01, -7.9077e-02],
                        [-7.2631e-02, -8.2833e-02, -7.4310e-02]],
              
                       [[ 4.9653e-02,  6.4736e-02,  4.3042e-02],
                        [ 6.7105e-02,  1.3038e-01,  8.5897e-02],
                        [ 6.0010e-02,  9.3700e-02,  7.2056e-02]]],
              
              
                      [[[ 4.6965e-02,  4.4167e-02,  5.0253e-02],
                        [-4.1048e-04,  2.4404e-04,  3.5650e-04],
                        [-4.6780e-02, -4.1869e-02, -5.0713e-02]],
              
                       [[ 5.3582e-02,  6.0488e-02,  6.5059e-02],
                        [-6

In [18]:
import torchvision
print(torchvision.__version__)

0.17.1+cu121


In [19]:
try:
    import transformers
    print("Transformers library is installed. Version:", transformers.__version__)
except ImportError:
    print("Transformers library is not installed.")

Transformers library is installed. Version: 4.39.1


In [20]:
from transformers import ViTModel, ViTConfig

In [21]:
configuration = ViTConfig(hidden_size = 144)

In [22]:
model = ViTModel(configuration)

In [23]:
configuration = model.config

In [24]:
print(configuration)

ViTConfig {
  "attention_probs_dropout_prob": 0.0,
  "encoder_stride": 16,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.0,
  "hidden_size": 144,
  "image_size": 224,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "model_type": "vit",
  "num_attention_heads": 12,
  "num_channels": 3,
  "num_hidden_layers": 12,
  "patch_size": 16,
  "qkv_bias": true,
  "transformers_version": "4.39.1"
}



In [27]:
import tensorflow as tf
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense, MaxPooling2D, Conv2D, Flatten, \
    BatchNormalization, Activation, GlobalAveragePooling2D, DepthwiseConv2D, \
    Dropout, ReLU, Concatenate, Input, add, Conv1D, MaxPooling1D
from tensorflow.keras.callbacks import EarlyStopping

## Testing Images

In [29]:
import glob
import cv2
import time

# Record the starting time
start_time = time.time()

test_image = cv2.imread("/home/t326h379/AffectNet-8Labels/val_set/images/3362.jpg")

# Record the end time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = end_time - start_time

# Print or use the elapsed time as needed
print(f"Execution time: {elapsed_time} seconds")

Execution time: 0.3089725971221924 seconds


In [30]:
import numpy as np

# Record the starting time
start_time = time.time()

val_images_images = np.array(test_image)

# Record the end time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = end_time - start_time

# Print or use the elapsed time as needed
print(f"Execution time: {elapsed_time} seconds")

Execution time: 0.0002512931823730469 seconds


In [31]:
type(test_image)

numpy.ndarray

In [32]:
test_image.shape

(224, 224, 3)

In [33]:
import glob
val_images = [file.split("/")[-1] for file in glob.glob("/home/t326h379/AffectNet-8Labels/val_set/images/*.jpg")]

In [34]:
val_images

['1990.jpg',
 '1245.jpg',
 '4969.jpg',
 '302.jpg',
 '763.jpg',
 '1183.jpg',
 '1624.jpg',
 '63.jpg',
 '2731.jpg',
 '3607.jpg',
 '2047.jpg',
 '1152.jpg',
 '2854.jpg',
 '2426.jpg',
 '474.jpg',
 '806.jpg',
 '1533.jpg',
 '1294.jpg',
 '4754.jpg',
 '4492.jpg',
 '3499.jpg',
 '3448.jpg',
 '2208.jpg',
 '4443.jpg',
 '4831.jpg',
 '5203.jpg',
 '2669.jpg',
 '4022.jpg',
 '1239.jpg',
 '4915.jpg',
 '5480.jpg',
 '4106.jpg',
 '1658.jpg',
 '1689.jpg',
 '5451.jpg',
 '4211.jpg',
 '2828.jpg',
 '3084.jpg',
 '3723.jpg',
 '1076.jpg',
 '4728.jpg',
 '131.jpg',
 '696.jpg',
 '5168.jpg',
 '2502.jpg',
 '2970.jpg',
 '3897.jpg',
 '922.jpg',
 '550.jpg',
 '4349.jpg',
 '1865.jpg',
 '1417.jpg',
 '1361.jpg',
 '581.jpg',
 '226.jpg',
 '3846.jpg',
 '3393.jpg',
 '2274.jpg',
 '647.jpg',
 '1700.jpg',
 '2615.jpg',
 '379.jpg',
 '5487.jpg',
 '5320.jpg',
 '1599.jpg',
 '4560.jpg',
 '4912.jpg',
 '4101.jpg',
 '718.jpg',
 '4677.jpg',
 '4216.jpg',
 '5456.jpg',
 '2164.jpg',
 '3083.jpg',
 '691.jpg',
 '136.jpg',
 '1071.jpg',
 '4088.jpg',
 '3

In [35]:
import glob
import os
os.chdir('/home/t326h379/AffectNet-8Labels/val_set/annotations/')
val_exp_file_count =[]
for name in glob.glob('*.npy'):
    a = name.split("_")
    if a[1].startswith("exp"):
        val_exp_file_count.append(name)

In [36]:
val_exp_file_count

['5366_exp.npy',
 '1623_exp.npy',
 '2867_exp.npy',
 '1469_exp.npy',
 '391_exp.npy',
 '2016_exp.npy',
 '4904_exp.npy',
 '4647_exp.npy',
 '5454_exp.npy',
 '246_exp.npy',
 '2724_exp.npy',
 '3198_exp.npy',
 '3537_exp.npy',
 '1499_exp.npy',
 '4185_exp.npy',
 '5396_exp.npy',
 '1036_exp.npy',
 '2603_exp.npy',
 '1847_exp.npy',
 '2449_exp.npy',
 '361_exp.npy',
 '4218_exp.npy',
 '1704_exp.npy',
 '2940_exp.npy',
 '5241_exp.npy',
 '4823_exp.npy',
 '453_exp.npy',
 '3168_exp.npy',
 '1990_exp.npy',
 '1843_exp.npy',
 '365_exp.npy',
 '580_exp.npy',
 '3414_exp.npy',
 '5392_exp.npy',
 '1278_exp.npy',
 '4764_exp.npy',
 '1032_exp.npy',
 '4181_exp.npy',
 '2893_exp.npy',
 '1994_exp.npy',
 '3326_exp.npy',
 '2135_exp.npy',
 '4827_exp.npy',
 '5245_exp.npy',
 '1700_exp.npy',
 '2944_exp.npy',
 '2258_exp.npy',
 '570_exp.npy',
 '2012_exp.npy',
 '395_exp.npy',
 '3201_exp.npy',
 '1627_exp.npy',
 '4171_exp.npy',
 '5128_exp.npy',
 '4794_exp.npy',
 '5362_exp.npy',
 '3533_exp.npy',
 '3779_exp.npy',
 '242_exp.npy',
 '2720

In [37]:
annotation_files_number_val = set()
for i in range(len(val_exp_file_count)):
    val_exp = val_exp_file_count[i].split("_")
    print(val_exp[0])
    annotation_files_number_val.add(val_exp[0])

5366
1623
2867
1469
391
2016
4904
4647
5454
246
2724
3198
3537
1499
4185
5396
1036
2603
1847
2449
361
4218
1704
2940
5241
4823
453
3168
1990
1843
365
580
3414
5392
1278
4764
1032
4181
2893
1994
3326
2135
4827
5245
1700
2944
2258
570
2012
395
3201
1627
4171
5128
4794
5362
3533
3779
242
2720
1964
5450
1115
879
881
5007
1542
4214
2592
3164
2377
615
4526
4189
2445
588
3656
1357
871
3771
3194
2728
4333
1465
4179
1280
5120
4908
3043
3209
3190
4405
875
3047
399
2254
1284
4798
5124
2139
2596
3160
1998
1546
5003
885
5249
3652
123
1491
4768
1274
5341
3819
4318
4152
1604
2840
3222
2031
4923
5296
4660
1136
1599
5473
2997
1947
2549
1011
4747
3437
3098
2624
1860
2967
4075
5483
1569
4690
5266
474
4804
2116
291
3305
2620
1864
3433
3679
4743
979
4509
295
3301
2358
2112
4800
5028
5262
1388
4071
1727
2963
5487
3689
2035
557
1894
3226
1600
2707
1943
265
5477
26
2993
4081
5292
1132
4664
1380
3932
5020
1565
4079
3143
3309
632
4808
478
971
2628
100
3094
2462
3671
2287
4089
1595
4426
1370
3756
2545
4314
1442


In [38]:
import os
import pickle
os.chdir('/home/t326h379/AffectNet-8Labels/val_set/annotations/')
with open('Kritika_AffectNet_Val.pickle', 'rb') as handle:
    a = pickle.load(handle)

In [39]:
y_val = []
for i in range(len(val_images)):
    y_val.append(int(a[val_images[i].split(".")[0]]))

In [40]:
y_val

[0,
 2,
 6,
 7,
 1,
 5,
 0,
 1,
 6,
 6,
 6,
 4,
 5,
 4,
 5,
 3,
 2,
 4,
 1,
 5,
 6,
 4,
 1,
 7,
 7,
 7,
 0,
 1,
 7,
 6,
 4,
 5,
 7,
 7,
 7,
 6,
 7,
 7,
 2,
 7,
 5,
 5,
 0,
 4,
 7,
 1,
 3,
 2,
 3,
 4,
 3,
 1,
 3,
 1,
 3,
 1,
 5,
 7,
 3,
 1,
 1,
 2,
 5,
 4,
 2,
 0,
 7,
 7,
 6,
 7,
 3,
 5,
 1,
 1,
 4,
 1,
 5,
 4,
 7,
 6,
 0,
 3,
 6,
 3,
 1,
 3,
 3,
 2,
 6,
 3,
 6,
 0,
 0,
 6,
 5,
 0,
 1,
 3,
 5,
 4,
 1,
 0,
 3,
 0,
 0,
 5,
 5,
 5,
 0,
 4,
 6,
 6,
 3,
 2,
 5,
 3,
 0,
 7,
 5,
 5,
 4,
 5,
 6,
 3,
 3,
 1,
 5,
 3,
 1,
 0,
 1,
 7,
 5,
 0,
 5,
 4,
 1,
 1,
 7,
 4,
 7,
 5,
 3,
 7,
 7,
 7,
 6,
 6,
 3,
 6,
 6,
 5,
 2,
 4,
 6,
 6,
 5,
 4,
 5,
 5,
 6,
 2,
 3,
 2,
 6,
 1,
 3,
 3,
 0,
 0,
 7,
 2,
 1,
 0,
 2,
 4,
 3,
 6,
 3,
 0,
 4,
 5,
 5,
 2,
 1,
 3,
 2,
 4,
 0,
 7,
 1,
 7,
 6,
 0,
 2,
 0,
 0,
 1,
 0,
 1,
 2,
 2,
 7,
 5,
 4,
 1,
 4,
 6,
 1,
 4,
 6,
 5,
 6,
 4,
 0,
 4,
 5,
 0,
 0,
 5,
 2,
 1,
 3,
 4,
 3,
 7,
 1,
 5,
 7,
 5,
 5,
 2,
 1,
 5,
 0,
 6,
 6,
 2,
 5,
 6,
 5,
 5,
 4,
 7,
 1,
 5,
 3,
 7,
 2,
 6,


In [41]:
y_val = np.array(y_val)

In [42]:
from tensorflow.keras.utils import to_categorical
num_classes = 8 
y_val_one_hot = to_categorical(y_val, num_classes)

In [43]:
val_images_images.shape

(224, 224, 3)

In [44]:
y_val.shape

(3999,)

In [45]:
val_images = np.array(val_images)

In [46]:
val_images.shape

(3999,)

## Training Images

In [47]:
import glob
import cv2
import time

# Record the starting time
start_time = time.time()

images = [cv2.imread(file) for file in glob.glob("/home/t326h379/AffectNet-8Labels/train_set/train_set/images/*.jpg")]

# Record the end time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = end_time - start_time

# Print or use the elapsed time as needed
print(f"Execution time: {elapsed_time} seconds")

Execution time: 4725.6852350234985 seconds


# It took me 4725.685 (78.75 minutes) no of seconds to load the training dataset

In [48]:
len(images)

287651

In [49]:
import numpy as np

# Record the starting time
start_time = time.time()

train_images = np.array(images)

# Record the end time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = end_time - start_time

# Print or use the elapsed time as needed
print(f"Execution time: {elapsed_time} seconds")

Execution time: 22.847601890563965 seconds


In [50]:
len(images)

287651

In [51]:
globe_loaded_images = [file.split("/")[-1] for file in glob.glob("/home/t326h379/AffectNet-8Labels/train_set/train_set/images/*.jpg")]

In [1]:
import glob
import os
os.chdir('/home/t326h379/AffectNet-8Labels/train_set/train_set/annotations/')
training_exp_file_count =[]
for name in glob.glob('*.npy'):
    a = name.split("_")
    if a[1].startswith("exp"):
        training_exp_file_count.append(name)

In [2]:
training_exp_file_count

['111968_exp.npy',
 '379238_exp.npy',
 '236864_exp.npy',
 '115882_exp.npy',
 '667_exp.npy',
 '344140_exp.npy',
 '408524_exp.npy',
 '302701_exp.npy',
 '371626_exp.npy',
 '397468_exp.npy',
 '280830_exp.npy',
 '177250_exp.npy',
 '246707_exp.npy',
 '28217_exp.npy',
 '208558_exp.npy',
 '49681_exp.npy',
 '80206_exp.npy',
 '301945_exp.npy',
 '169215_exp.npy',
 '372862_exp.npy',
 '199400_exp.npy',
 '258742_exp.npy',
 '36252_exp.npy',
 '273061_exp.npy',
 '108683_exp.npy',
 '376988_exp.npy',
 '200146_exp.npy',
 '131411_exp.npy',
 '88618_exp.npy',
 '283674_exp.npy',
 '142536_exp.npy',
 '218811_exp.npy',
 '320280_exp.npy',
 '17997_exp.npy',
 '121958_exp.npy',
 '413237_exp.npy',
 '349208_exp.npy',
 '275973_exp.npy',
 '189949_exp.npy',
 '401786_exp.npy',
 '271899_exp.npy',
 '307057_exp.npy',
 '374170_exp.npy',
 '331975_exp.npy',
 '298167_exp.npy',
 '85550_exp.npy',
 '342852_exp.npy',
 '159225_exp.npy',
 '75345_exp.npy',
 '268772_exp.npy',
 '113190_exp.npy',
 '194748_exp.npy',
 '399964_exp.npy',
 '24

In [8]:
pwd

'/homes/t326h379/AffectNet-8Labels/train_set/train_set/annotations'

In [10]:
import numpy as np
np.load(training_exp_file_count[0])

array('3', dtype='<U1')

In [3]:
annotation_files_number_training = set()
for i in range(len(training_exp_file_count)):
    train_exp = training_exp_file_count[i].split("_")
    annotation_files_number_training.add(train_exp[0])

In [4]:
annotation_files_number_training

{'77903',
 '287483',
 '6286',
 '388704',
 '330304',
 '369438',
 '251590',
 '384043',
 '150688',
 '29592',
 '290025',
 '23839',
 '353287',
 '311023',
 '25451',
 '74834',
 '260737',
 '192421',
 '234668',
 '362909',
 '381060',
 '183118',
 '113090',
 '321431',
 '189383',
 '71033',
 '135089',
 '82650',
 '237038',
 '234123',
 '60595',
 '359794',
 '134672',
 '365247',
 '50254',
 '189972',
 '136244',
 '35430',
 '347116',
 '157265',
 '60851',
 '380534',
 '159959',
 '229757',
 '163999',
 '297302',
 '301721',
 '148958',
 '409275',
 '84339',
 '302322',
 '172792',
 '97902',
 '345808',
 '155301',
 '284960',
 '178470',
 '186156',
 '54884',
 '377325',
 '195222',
 '67296',
 '15009',
 '271367',
 '34877',
 '87865',
 '329816',
 '147795',
 '100112',
 '14027',
 '276624',
 '377934',
 '211532',
 '49427',
 '229844',
 '30333',
 '305604',
 '205978',
 '269805',
 '7291',
 '110382',
 '175334',
 '48854',
 '310468',
 '5363',
 '156375',
 '57559',
 '250157',
 '297706',
 '317864',
 '400263',
 '32359',
 '336400',
 '27460

In [11]:
import pickle
with open('Kritika_AffectNet_Training.pickle', 'rb') as handle:
    b = pickle.load(handle)

In [15]:
type(b)

dict

In [21]:
keys_of_all_images = []
values_of_all_images = []
for key, value in b.items():
    keys_of_all_images.append(key+".jpg") 
    values_of_all_images.append(int(value))


In [25]:
import pandas as pd
data = {
    'Image': keys_of_all_images,
    'Label': values_of_all_images
}

# Create a DataFrame
df = pd.DataFrame(data)

# Display the DataFrame

In [27]:
df.to_csv("Image_to_lable.csv",index=False)

In [28]:
pwd

'/homes/t326h379/AffectNet-8Labels/train_set/train_set/annotations'

In [55]:
y_train = []
for i in range(len(globe_loaded_images)):
    y_train.append(int(b[globe_loaded_images[i].split(".")[0]]))

In [56]:
y_train = np.array(y_train)

In [57]:
from tensorflow.keras.utils import to_categorical
# One-hot encode the target labels
num_classes = 8  # Replace with the actual number of classes
y_train_one_hot = to_categorical(y_train, num_classes)

In [58]:
y_train.shape

(287651,)

In [59]:
y_train_one_hot.shape

(287651, 8)

In [60]:
# hf_model(tf.expand_dims(test_image, axis = 0))

In [61]:
# hf_model.summary()

In [62]:
import tensorflow as tf
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense, MaxPooling2D, Conv2D, Flatten, \
    BatchNormalization, Activation, GlobalAveragePooling2D, DepthwiseConv2D, \
    Dropout, ReLU, Concatenate, Input, add, Conv1D, MaxPooling1D
from tensorflow.keras.callbacks import EarlyStopping
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets

In [63]:
X_train, X_test, y_train, y_test = train_test_split(train_images, y_train_one_hot, stratify= y_train, test_size=0.1, random_state=42)

In [64]:
X_train_80, X_test_10, y_train_80, y_test_10 = train_test_split(X_train, y_train, stratify= y_train, test_size=0.1, random_state=42)

In [65]:
print(X_train_80.shape, X_test_10.shape, y_train_80.shape, y_test_10.shape)

(232996, 224, 224, 3) (25889, 224, 224, 3) (232996, 8) (25889, 8)


In [66]:
type(y_train_80)

numpy.ndarray

In [67]:
y_train_80 = np.argmax(y_train_80, axis=1)

In [68]:
y_train_80

array([6, 1, 6, ..., 2, 1, 3])

In [69]:
y_test_10 = np.argmax(y_test_10, axis=1)

In [70]:
y_test_10

array([2, 0, 4, ..., 2, 4, 1])

In [71]:
# Convert the image to 32-bit floating-point format
X_train_80_float32 = X_train_80.astype(np.float32) / 255.0
X_test_10_float32 = X_test_10.astype(np.float32) / 255.0

In [72]:
import torch
from torch.utils.data import DataLoader, TensorDataset

# Convert numpy arrays to PyTorch tensors
X_train_tensor = torch.tensor(X_train_80_float32, dtype=torch.float32)
X_train_tensor = X_train_tensor.permute(0, 3, 1, 2)
y_train_tensor = torch.tensor(y_train_80)

X_val_tensor = torch.tensor(X_test_10_float32, dtype=torch.float32)
X_val_tensor = X_val_tensor.permute(0, 3, 1, 2)
y_val_tensor = torch.tensor(y_test_10)

# Create datasets
train_dataset = TensorDataset(X_train_tensor, y_train_tensor)
val_dataset = TensorDataset(X_val_tensor, y_val_tensor)

# Create dataloaders
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=32, shuffle=False)

In [73]:
X_train_tensor.shape

torch.Size([232996, 3, 224, 224])

In [74]:
y_train_tensor.shape

torch.Size([232996])

In [75]:
X_val_tensor.shape

torch.Size([25889, 3, 224, 224])

In [76]:
y_val_tensor.shape

torch.Size([25889])

In [82]:
import torch
from torchvision import transforms
from PIL import Image
from fastervit import create_model

# Define fastervit-0 model with 224 x 224 resolution

# Step 1: Load the pre-trained FasterViT model
model = create_model('faster_vit_0_224', 
                          pretrained=True,
                          model_path="/tmp/faster_vit_0.pth.tar")
model.eval()

num_ftrs = model.head.in_features
model.head = torch.nn.Linear(num_ftrs, 8)

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
model = model.to(device)

In [83]:
model

FasterViT(
  (patch_embed): PatchEmbed(
    (proj): Identity()
    (conv_down): Sequential(
      (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (1): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (2): ReLU()
      (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (4): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (5): ReLU()
    )
  )
  (levels): ModuleList(
    (0): FasterViTLayer(
      (blocks): ModuleList(
        (0): ConvBlock(
          (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          (act1): GELU(approximate='none')
          (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_s

In [84]:
import torch.optim as optim
import torch.nn as nn

criterion = nn.CrossEntropyLoss()  # or any other suitable loss function
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Started from 6/25/2024 5:48 PM

In [85]:
num_epochs = 12  # Set the number of epochs

for epoch in range(num_epochs):
    model.train()  # Set the model to training mode
    running_loss = 0.0
    print(epoch)

    for images, labels in train_loader:
        optimizer.zero_grad()  # Zero the parameter gradients

        outputs = model(images)  # Forward pass
        loss = criterion(outputs, labels)  # Compute the loss
        loss.backward()  # Backward pass
        optimizer.step()  # Optimize the parameters

        running_loss += loss.item()

    # Print training statistics
    print(f"Epoch {epoch+1}/{num_epochs}, Loss: {running_loss/len(train_loader)}")

    # Validation phase
    model.eval()  # Set the model to evaluation mode
    val_loss = 0.0
    correct = 0
    total = 0

    with torch.no_grad():
        for images, labels in val_loader:
            outputs = model(images)
            loss = criterion(outputs, labels)
            val_loss += loss.item()

            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    # Print validation statistics
    print(f"Validation Loss: {val_loss/len(val_loader)}, Accuracy: {100 * correct / total}%")

0
Epoch 1/12, Loss: 0.8544305484395078
Validation Loss: 0.7482870716003724, Accuracy: 73.78423268569662%
1
Epoch 2/12, Loss: 0.7458240338298063
Validation Loss: 0.7090106909657702, Accuracy: 74.83100930897292%
2
Epoch 3/12, Loss: 0.7073500329765178
Validation Loss: 0.7197658028131649, Accuracy: 74.93916335123025%
3
Epoch 4/12, Loss: 0.6782397664582065
Validation Loss: 0.6913847922542948, Accuracy: 75.15547143574491%
4
Epoch 5/12, Loss: 0.6531706755324498
Validation Loss: 0.6659632724744302, Accuracy: 76.26405036888254%
5
Epoch 6/12, Loss: 0.6275596510218874
Validation Loss: 0.6690071006615956, Accuracy: 76.21769863648655%
6
Epoch 7/12, Loss: 0.6017875128625939
Validation Loss: 0.6718993489205102, Accuracy: 76.29881416817953%
7
Epoch 8/12, Loss: 0.5746987965924829
Validation Loss: 0.6751330115728909, Accuracy: 75.93186295337789%
8
Epoch 9/12, Loss: 0.5433206453313706
Validation Loss: 0.6903783945021806, Accuracy: 76.15203368225887%
9
Epoch 10/12, Loss: 0.510098680372963
Validation Loss:

In [86]:
len(train_loader)

7282

In [87]:
torch.save(model.state_dict(), 'Twelve_Iteration_fastervit_model.pth')

# Independent Test Dataset Testing

In [88]:
print(X_test.shape, y_test.shape)

(28766, 224, 224, 3) (28766, 8)


In [89]:
X_test_float32 = X_test.astype(np.float32) / 255.0

In [90]:
y_test_integer = np.argmax(y_test, axis=1)

In [91]:
y_test_integer.shape

(28766,)

In [92]:
import torch
from torch.utils.data import DataLoader, TensorDataset

# Convert numpy arrays to PyTorch tensors
X_test_tensor = torch.tensor(X_test_float32, dtype=torch.float32)
X_test_tensor = X_test_tensor.permute(0, 3, 1, 2)
y_test_tensor = torch.tensor(y_test_integer)

In [93]:
y_test_tensor.shape

torch.Size([28766])

In [94]:
X_test_tensor.shape

torch.Size([28766, 3, 224, 224])

In [95]:
# Create datasets
test_dataset = TensorDataset(X_test_tensor, y_test_tensor)

# Create dataloaders
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)

In [96]:
model.eval()

FasterViT(
  (patch_embed): PatchEmbed(
    (proj): Identity()
    (conv_down): Sequential(
      (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (1): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (2): ReLU()
      (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (4): BatchNorm2d(64, eps=0.0001, momentum=0.1, affine=True, track_running_stats=True)
      (5): ReLU()
    )
  )
  (levels): ModuleList(
    (0): FasterViTLayer(
      (blocks): ModuleList(
        (0): ConvBlock(
          (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          (act1): GELU(approximate='none')
          (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          (norm2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_s

In [97]:
# Define a loss function and other necessary components
criterion = torch.nn.CrossEntropyLoss()

# Testing loop
test_loss = 0.0
correct = 0
total = 0

with torch.no_grad():  # No need to calculate gradients during inference
    for data, targets in test_loader:
        outputs = model(data)
        loss = criterion(outputs, targets)
        test_loss += loss.item() * data.size(0)
        _, predicted = torch.max(outputs, 1)
        total += targets.size(0)
        correct += (predicted == targets).sum().item()

# Calculate average loss and accuracy
test_loss = test_loss / len(test_loader.dataset)
accuracy = 100 * correct / total

print(f'Test Loss: {test_loss:.4f}, Test Accuracy: {accuracy:.2f}%')

Test Loss: 0.8027, Test Accuracy: 74.80%


In [98]:
# Collect true and predicted labels
all_labels = []
all_preds = []

with torch.no_grad():  # No need to calculate gradients during inference
    for data, targets in test_loader:
        outputs = model(data)
        _, predicted = torch.max(outputs, 1)
        all_labels.extend(targets.cpu().numpy())
        all_preds.extend(predicted.cpu().numpy())

In [99]:
# Compute MCC
from sklearn.metrics import matthews_corrcoef

mcc = matthews_corrcoef(all_labels, all_preds)
print(f'Matthews Correlation Coefficient: {mcc:.4f}')

Matthews Correlation Coefficient: 0.6321


In [100]:
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay

# Compute confusion matrix
cm = confusion_matrix(all_labels, all_preds)

cm

array([[ 5454,   888,   292,   308,    56,    12,   460,    18],
       [  844, 12278,    41,   162,    15,    11,    70,    21],
       [  854,   109,  1188,    50,    82,    14,   248,     1],
       [  273,   224,    23,   723,   121,     0,    44,     1],
       [   74,    38,    54,   126,   294,     7,    45,     0],
       [   62,    46,    29,     9,    20,    87,   125,     2],
       [  645,    97,   110,    60,    58,    39,  1475,     4],
       [  112,   215,     1,     6,     0,     1,    23,    17]])

In [1]:
pip list

Package                           Version      Editable project location
--------------------------------- ------------ -------------------------
absl-py                           2.1.0
accelerate                        0.30.1
aiofiles                          23.2.1
aiohttp                           3.9.3
aiosignal                         1.3.1
alabaster                         0.7.12
annotated-types                   0.7.0
anyio                             3.6.1
appdirs                           1.4.4
argon2-cffi                       20.1.0
asn1crypto                        1.5.1
asttokens                         2.0.8
astunparse                        1.6.3
async-generator                   1.10
async-timeout                     4.0.3
atomicwrites                      1.4.0
attrs                             23.2.0
audioread                         3.0.1
av                                14.1.0
Babel                             2.10.1
backcall                          0.2.0
backport

Note: you may need to restart the kernel to use updated packages.


# Thank You